core/src/database/PostgreSQLBackend.cpp

```diff
@@ -32,17 +32,26 @@
 #include <utils/Exception.hpp>
 #include <api/ConfigurationDefaults.hpp>
 #include <soci-postgresql.h>
+
+#ifdef SSL_SUPPORT
+    #include <openssl/ssl.h>
+#endif
+
 using namespace soci;
 
 namespace ledger {
     namespace core {
         PostgreSQLBackend::PostgreSQLBackend() : DatabaseBackend(),
                                                  _connectionPoolSize(api::ConfigurationDefaults::DEFAULT_PG_CONNECTION_POOL_SIZE)
-        {}
+        {
+            initSSLLibraries();
```

Comment:
The main problem with going this way is that you’re doing something non-trivial — initializing OpenSSL — in something that is not immediately related (i.e. a backend database).

I’m not sure this is the right place to do that.

Reply:
I think the _saner_ approach would be to be that initializing in the `WalletPool` (which will translate to `Services` in the modularized core) and ask for its initialization there. That means that we cannot create a `PostgreSQLBackend` with the default ctor — that seems perfectly fine to me, as it has an external dependency.

What do you guys think? It could be something like:

```cpp
WalletPool::initSSL() {
  static done = false;
  
  if (!done) {
#ifdef SSL_SUPPORT
    #if OPENSSL_VERSION_NUMBER < 0x10100000L
            SSL_library_init();
    #else
            OPENSSL_init_ssl(0, NULL);
    #endif
#endif

    done = true;
  }
}
```

Reply:
Thought about it ! But potentially we could use PG without WalletPool in a future, then it would take us time time to think about that, we should bind PG with SSL support.

Reply:
Second point, I don't like the fact that client should call that explicitly ! They should not worry about that

Reply:
https://en.cppreference.com/w/cpp/thread/call_once is thread safe variant of this.

Reply:
`call_once` is a really good point! @KhalilBellakrid that’s I split `WalletPool` into `Services + WalletStore` in the modularization. :)

Reply:
@teams2ua as discussed, agreed ! 👌 
@phaazon that's exactly kind of things to enhance in new Core, so I let you port that fix on it taking into consideration all what have been discussed above ! Thank you sir !